### PR TITLE
Rollup class fields support 

### PR DESCRIPTION
**Pull request by guybedford**
This adds class fields and static class fields support.

The reason being that it is almost impossible to add via the inject mechanism due to the different Acorn instances in play giving this error - https://github.com/acornjs/acorn-private-class-elements/blob/master/index.js#L23.

In addition both plugins needed to be forked to add the Acorn 7 support which is missing (with other similar PRs outstanding for some time).

For me personally this was to get the feature working, but I suspect there may be components of the treeshaking that will need attention. In addition the tests have not been fleshed out, hence the WIP status. But this is functional in my own test cases.

Maybe RollupJS can distribute some of its funds to help Acorn plugin development. @adrianheine would financial support help here in moving things forward?

This PR contains:
- [ ] bugfix
- [x] feature
- [ ] refactor
- [ ] documentation
- [ ] other

Are tests included?
- [ ] yes (*bugfixes and features will not be merged without tests*)
- [x] no

Breaking Changes?
- [ ] yes (*breaking changes will not be merged unless absolutely necessary*)
- [x] no

List any relevant issue numbers:

### CODE REVIEW

File: package.json [hunk 1]
```diff
@@ -143,5 +143,9 @@
       "default": "./dist/rollup.js"
     },
     "./dist/": "./dist/"
+  },
+  "dependencies": {
+    "acorn-class-fields": "github:guybedford/acorn-class-fields#acorn-update",
+    "acorn-static-class-features": "github:guybedford/acorn-static-class-features#patch-1"
```
Comment (guybedford) [inline package.json:149]: These should be `devDependencies` of course.

**Comment by lukastaegert** [comment]
On second thought, maybe I want to at least add the new AST node types for class properties...

**Comment by lukastaegert** [comment]
Ok, Nodes are added.

**Comment by guybedford** [comment]
I do not pretend to understand this magic, but it looks great! I suppose we can remove the WIP status of this PR then?

Should we test static expressions and static functions as well here?

**Comment by guybedford** [comment]
*private static methods I mean

**Comment by lukastaegert** [comment]
> I suppose we can remove the WIP status of this PR then?

Definitely 

> Should we test static expressions and static functions as well here?

Please feel free to add this, I must admit I was not entirely sure what was included. I will not be able to look at this again before tonight.

**Comment by lukastaegert** [comment]
@guybedford I extended the test to include all new features and, who would have thought, support is incomplete: Private methods, setters and getters produce an "unexpected token" while parsing. I assume this needs to be fixed in the plugin. Should we still release it as "incomplete support"? It still provides some value and also fixes a bug.

**Comment by lukastaegert** [comment]
You can remove the comments in my test case to see what I mean.

**Comment by guybedford** [comment]
@lukastaegert thankfully, for now at least, I believe this syntax support level exactly matches Chrome stable (and hence the major usage). But yes as soon as Chrome lands those that will be an issue. Perhaps maintaining and developing the forks as part of Rollup will be an option here.

Note that also eg Terser doesn't support any of this yet (I know @fabiosantoscode was originally against it)... perhaps this is another area where RollupJS could assist with funding for these features.

**Comment by lukastaegert** [comment]
You are right, Chrome does not support any of the missing ones. It also does not support static private methods, so there is at least one additional case that is supported. Then I will push this one out as well.

**Comment by bathos** [comment]
Private methods and accessors, both static and instance, are implemented in V8 and are unflagged now in Chrome 84 as well. Will Rollup be supporting these as well, and if so, should an issue be opened to track that?

**Comment by bathos** [comment]
I doubt it’s really this simple, but FWIW, in this code —

```js
    // Parse private static methods
    parsePropertyName(prop) {
      if (prop.static && this.type == this.privateNameToken) {
        this.parsePrivateClassElementName(prop);
      } else {
        super.parsePropertyName(prop);
      }
    }
```

If I remove `prop.static` from the condition, unexpected token errors from instance private methods stop and the output looks correct.

**Comment by lukastaegert** [comment]
You can create an issue, but this is not Rollup code, without an upstream issue in the corresponding acorn plugin (likely https://github.com/acornjs/acorn-private-class-elements), nothing will move forward here.

**Comment by bathos** [comment]
Cool — in that case, it seems like maybe Rollup support (at least for instance methods) might end up being “free” if the upstream plugin is updated to handle these cases. Will investigate a bit further, thanks.

**Comment by lukastaegert** [comment]
Yes, that is my expectation. Others will profit from this as well.

**Comment by fabiosantoscode** [comment]
Hey there!

I was against class fields, but I wouldn't stop progress for just that :) they are implemented in Terser.

**Pull request merged**